cloudinit/config/cc_ansible.py

```diff
@@ -0,0 +1,108 @@
+
+from cloudinit.cloud import Cloud
+from cloudinit.config.schema import MetaSchema, get_meta_doc
+from cloudinit.distros import ALL_DISTROS, Distro
+from cloudinit.settings import PER_INSTANCE
+from cloudinit.subp import ProcessExecutionError, subp, which
+
+meta: MetaSchema = {
+    "id": "cc_ansible",
+    "name": "Ansible",
+    "title": "Configure ansible for instance",
+    "description": dedent(
+        """\
+        This module provides ansible integration.``
+
+        Ansible is often used agentless and in parallel
+        across multiple hosts simultaneously. This
+        doesn't fit the model of cloud-init: a single
+        host configuring itself during boot. Instead,
+        this module installs ansible during boot and
+        then uses ``ansible-pull`` to run the playbook
+        repository at the remote URL.
+        """
+    ),
+    "distros": [ALL_DISTROS],
+    "examples": [
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              install: true
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+    ],
+    "frequency": PER_INSTANCE,
+}
+
+__doc__ = get_meta_doc(meta)
+
+
+def handle(name: str, cfg: dict, cloud: Cloud, log: Logger, _):
+    install, ansible_config = get_and_validate_config(cfg)
+    install_ansible(cloud.distro, install)
+    run_ansible_pull(ansible_config, log)
+
+
+def get_and_validate_config(cfg: dict) -> Tuple[bool, dict]:
+    ansible_cfg: dict = cfg.get("ansible", {})
+    pull: dict = ansible_cfg.get("pull", {})
+    install: bool = ansible_cfg.get("install", False)
+    if not all([pull.get("playbook-name"), pull.get("url")]):
+        raise ValueError(
+            "Missing required key: playbook-name and "
+            "url keys required for ansible module")
+    return (
+        install,
+        pull,
+    )
+
+
+def install_ansible(distro: Distro, install: bool):
+    """Give users flexibility in whether to use the package install module or
+    this module.
+    """
+    if install:
+        distro.install_packages("ansible")
+
+
+def check_deps(dep: str):
+    if not which(dep):
+        raise ValueError(
+            f"command: {dep} is not available, please set"
+            "ansible.install: True in your config or otherwise ensure that"
+            "it is installed (either in your base image or in a package"
+            "install module)"
+        )
+
+
+def run_ansible_pull(cfg: dict, log: Logger):
+    cmd = "ansible-pull"
+    check_deps(cmd)
+    playbook_name: str = cfg.pop("playbook-name")
+    try:
+
+        stdout, stderr = subp(
+            [
+                cmd,
+                *[
+                    f"--{key}={value}" if value else f"--{key}"
+                    for key, value in cfg.items()
+                ],
+                playbook_name,
+            ]
+        )
+        if stderr:
+            log.warn(f"{stderr}")
```

Comment:
I think `/var/log/cloud-init-output.log` is the right place to stuff this output?

Reply:
not sure why we'd see stderr here that didn't raise a ProcessExecutionError. If we do have known password/credentials leaks in messages emitted by ansible-pull I agree this should be steered to cloud-init-output.log to avoid the world-readable cloud-init.log if we can think of an easy way to do that.

Reply:
> not sure why we'd see stderr here that didn't raise a ProcessExecutionError

+1

> If we do have known password/credentials leaks in messages emitted by ansible-pull 

I don't think ansible typically logs secrets to stdout, my concern is that this will log arbitrary ansible task descriptions which _could_ contain sensitive data. I don't feel strongly about this point, it's more of a hand wavy "lets not surprise users that may expect ansible logs to be root-only readable".

Perhaps a better argument for logging to `/etc/cloud/cloud-init-out.log` is that the output[1] is multiline, which I think fits better in `/etc/cloud/cloud-init-out.log`

[1] output example:
```
Starting Ansible Pull at 2022-07-22 04:09:06
/usr/bin/ansible-pull --url=https://github.com/holmanb/vmboot.git ubuntu.yml
[WARNING]: Could not match supplied host pattern, ignoring: me.lxd
[WARNING]: Could not match supplied host pattern, ignoring: me
localhost | CHANGED => {
    "after": "c5d28021d861cecf76905d1c2bc6ceb366c4ce95",
    "before": null,
    "changed": true
}
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
[WARNING]: Could not match supplied host pattern, ignoring: me.lxd
[WARNING]: Could not match supplied host pattern, ignoring: me

PLAY [127.0.0.1] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [apt : install packages] **************************************************
ok: [localhost]

TASK [apt : install packages] **************************************************
changed: [localhost] => (item=git)

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```